### PR TITLE
statistics: fix data race in the Handle.mu.ctx

### LINKED ISSUE #34715: DATA RACE with loadStatsWorker and UpdateTableStatsLoop
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c0044c3538 by goroutine 144:
  github.com/pingcap/tidb/sessionctx/variable.setSnapshotTS()
      /go/tidb/sessionctx/variable/varsutil.go:436 +0x90
  github.com/pingcap/tidb/sessionctx/variable.glob..func21()
      /go/tidb/sessionctx/variable/sysvar.go:131 +0x44
  github.com/pingcap/tidb/sessionctx/variable.(*SysVar).SetSessionFromHook()
      /go/tidb/sessionctx/variable/variable.go:192 +0xb5
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).SetSystemVar()
      /go/tidb/sessionctx/variable/session.go:1640 +0x59
  github.com/pingcap/tidb/session.(*session).useCurrentSession.func1()
      /go/tidb/session/session.go:1737 +0x14e
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:436 +0x32
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
      /go/tidb/session/session.go:1831 +0x124
  github.com/pingcap/tidb/statistics/handle.(*statsReader).read()
      /go/tidb/statistics/handle/handle.go:1444 +0x3b1
  github.com/pingcap/tidb/statistics/handle.(*Handle).fmSketchFromStorage()
      /go/tidb/statistics/handle/handle.go:743 +0x1af
  github.com/pingcap/tidb/statistics/handle.(*Handle).columnStatsFromStorage()
      /go/tidb/statistics/handle/handle.go:855 +0x1564
  github.com/pingcap/tidb/statistics/handle.(*Handle).TableStatsFromStorage()
      /go/tidb/statistics/handle/handle.go:942 +0x999
  github.com/pingcap/tidb/statistics/handle.(*Handle).Update()
      /go/tidb/statistics/handle/handle.go:332 +0x97c
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
      /go/tidb/domain/domain.go:1390 +0x725
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker-fm()
      <autogenerated>:1 +0x39
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Previous read at 0x00c0044c3538 by goroutine 143:
  github.com/pingcap/tidb/session.(*session).GetInfoSchema()
      /go/tidb/session/session.go:3444 +0x76
  github.com/pingcap/tidb/statistics/handle.(*Handle).updateGlobalStats()
      /go/tidb/statistics/handle/ddl.go:78 +0x9a
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleDDLEvent()
      /go/tidb/statistics/handle/ddl.go:58 +0x196
  github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()
      /go/tidb/domain/domain.go:1460 +0x4a8
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop.func1()
      /go/tidb/domain/domain.go:1330 +0x68
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Goroutine 144 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1319 +0x27c
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Goroutine 143 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1330 +0x4f6
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0044c34c8 by goroutine 144:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).SetStatusFlag()
      /go/tidb/sessionctx/variable/session.go:1481 +0x84
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).SetInTxn()
      /go/tidb/sessionctx/variable/session.go:1492 +0x4c
  github.com/pingcap/tidb/executor.(*SimpleExec).executeCommit()
      /go/tidb/executor/simple.go:690 +0x27
  github.com/pingcap/tidb/executor.(*SimpleExec).Next()
      /go/tidb/executor/simple.go:136 +0x44d
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:319 +0x5af
  github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor()
      /go/tidb/executor/adapter.go:665 +0x6a4
  github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay()
      /go/tidb/executor/adapter.go:516 +0x29c
  github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
      /go/tidb/executor/adapter.go:465 +0xfc9
  github.com/pingcap/tidb/session.runStmt()
      /go/tidb/session/session.go:2066 +0x6cb
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1935 +0xdba
  github.com/pingcap/tidb/session.(*session).ExecuteInternal()
      /go/tidb/session/session.go:1465 +0x464
  github.com/pingcap/tidb/statistics/handle.(*Handle).releaseStatsReader()
      /go/tidb/statistics/handle/handle.go:1495 +0xc2
  github.com/pingcap/tidb/statistics/handle.(*Handle).releaseGlobalStatsReader()
      /go/tidb/statistics/handle/handle.go:1466 +0xe4
  github.com/pingcap/tidb/statistics/handle.(*Handle).TableStatsFromStorage.func1()
      /go/tidb/statistics/handle/handle.go:902 +0x4b
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:436 +0x32
  github.com/pingcap/tidb/statistics/handle.(*Handle).Update()
      /go/tidb/statistics/handle/handle.go:332 +0x97c
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
      /go/tidb/domain/domain.go:1390 +0x725
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker-fm()
      <autogenerated>:1 +0x39
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Previous read at 0x00c0044c34c8 by goroutine 143:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).GetStatusFlag()
      /go/tidb/sessionctx/variable/session.go:1486 +0x30f
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InTxn()
      /go/tidb/sessionctx/variable/session.go:1500 +0x303
  github.com/pingcap/tidb/session.(*session).GetInfoSchema()
      /go/tidb/session/session.go:3447 +0x2da
  github.com/pingcap/tidb/statistics/handle.(*Handle).updateGlobalStats()
      /go/tidb/statistics/handle/ddl.go:78 +0x9a
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleDDLEvent()
      /go/tidb/statistics/handle/ddl.go:58 +0x196
  github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()
      /go/tidb/domain/domain.go:1460 +0x4a8
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop.func1()
      /go/tidb/domain/domain.go:1330 +0x68
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Goroutine 144 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1319 +0x27c
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Goroutine 143 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1330 +0x4f6
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0044c3478 by goroutine 144:
  github.com/pingcap/tidb/sessiontxn/legacy.(*SimpleTxnContextProvider).OnInitialize()
      /go/tidb/sessiontxn/legacy/provider.go:110 +0x49a
  github.com/pingcap/tidb/session.(*txnManager).EnterNewTxn()
      /go/tidb/session/txnmanager.go:82 +0x109
  github.com/pingcap/tidb/session.(*session).PrepareTxnCtx()
      /go/tidb/session/session.go:3167 +0x847
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1870 +0x279
  github.com/pingcap/tidb/session.(*session).ExecuteInternal()
      /go/tidb/session/session.go:1465 +0x464
  github.com/pingcap/tidb/statistics/handle.(*Handle).getStatsReader()
      /go/tidb/statistics/handle/handle.go:1484 +0x2ec
  github.com/pingcap/tidb/statistics/handle.(*Handle).getGlobalStatsReader()
      /go/tidb/statistics/handle/handle.go:1461 +0x124
  github.com/pingcap/tidb/statistics/handle.(*Handle).TableStatsFromStorage()
      /go/tidb/statistics/handle/handle.go:897 +0x84
  github.com/pingcap/tidb/statistics/handle.(*Handle).Update()
      /go/tidb/statistics/handle/handle.go:332 +0x97c
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
      /go/tidb/domain/domain.go:1390 +0x725
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker-fm()
      <autogenerated>:1 +0x39
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Previous read at 0x00c0044c3478 by goroutine 143:
  github.com/pingcap/tidb/session.(*session).GetInfoSchema()
      /go/tidb/session/session.go:3447 +0x2ef
  github.com/pingcap/tidb/statistics/handle.(*Handle).updateGlobalStats()
      /go/tidb/statistics/handle/ddl.go:78 +0x9a
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleDDLEvent()
      /go/tidb/statistics/handle/ddl.go:58 +0x196
  github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()
      /go/tidb/domain/domain.go:1460 +0x4a8
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop.func1()
      /go/tidb/domain/domain.go:1330 +0x68
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      /go/tidb/util/wait_group_wrapper.go:33 +0x73
Goroutine 144 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1319 +0x27c
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Goroutine 143 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      /go/tidb/util/wait_group_wrapper.go:31 +0xdc
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1330 +0x4f6
  github.com/pingcap/tidb/domain.(*Domain).LoadAndUpdateStatsLoop()
      /go/tidb/domain/domain.go:1299 +0x6b
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2939 +0xac4
  github.com/pingcap/tidb/ddl_test.(*stateChangeSuite).SetupSuite()
      /go/tidb/ddl/db_change_test.go:73 +0xf1
  github.com/stretchr/testify/suite.Run()
      /home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/suite/suite.go:128 +0x597
  github.com/pingcap/tidb/ddl_test.TestStateChange()
      /go/tidb/ddl/db_change_test.go:63 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
================== 
```

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #34715

Problem Summary:

We call (*Handle).updateGlobalStats, which uses Handle.mu.ctx to fetch another session from session pool to execute internal SQL. it will happen to a data race. 

### What is changed and how it works?

We can get sessionctx from sessionpool to avoid it.

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by tiancaiamao — APPROVED** [review]

**Comment by tiancaiamao** [comment]
I'm thinking about make build plan thread-safe.
Why building plan can't be parallel? because they share the same sessionctx and modify fields within them... 
but that reason is ridiculous and it's a taste of bad code, building plan is not executing, it should avoid modifying anything.

**Review by xuyifangreeneyes — APPROVED** [review]

File: statistics/handle/ddl.go [hunk 1]
```diff
@@ -75,7 +76,13 @@ var analyzeOptionDefault = map[ast.AnalyzeOptionType]uint64{
 func (h *Handle) updateGlobalStats(tblInfo *model.TableInfo) error {
 	// We need to merge the partition-level stats to global-stats when we drop table partition in dynamic mode.
 	tableID := tblInfo.ID
-	is := h.mu.ctx.GetInfoSchema().(infoschema.InfoSchema)
+	se, err := h.pool.Get()
+	if err != nil {
+		return err
+	}
+	sctx := se.(sessionctx.Context)
+	defer h.pool.Put(se)
```
Comment (xuyifangreeneyes) [inline statistics/handle/ddl.go:84]: Maybe we can put it back just after we get `InfoSchema`?

**Comment by xuyifangreeneyes** [comment]
> I'm thinking about make build plan thread-safe. Why building plan can't be parallel? because they share the same sessionctx and modify fields within them... but that reason is ridiculous and it's a taste of bad code, building plan is not executing, it should avoid modifying anything.

+1. We should avoid use `h.mu.ctx` as much as possible and maybe deprecate it in the future.

**Comment by hawkingrei** [comment]
+1

**Comment by hawkingrei** [comment]
> > I'm thinking about make build plan thread-safe. Why building plan can't be parallel? because they share the same sessionctx and modify fields within them... but that reason is ridiculous and it's a taste of bad code, building plan is not executing, it should avoid modifying anything.
> 
> +1. We should avoid use `h.mu.ctx` as much as possible and maybe deprecate it in the future.

We can use `h.pool` to replace with `h.mu.ctx`.

File: statistics/handle/ddl.go (hunk 1 above)
Comment (hawkingrei) [reply statistics/handle/ddl.go]: Done.

**Pull request merged**